### PR TITLE
fix(assembly): various formatting quirks

### PR DESCRIPTION
**Pull request by bitwalker**
This PR fixes a handful of bugs/quirks with the initial implementations of the PrettyPrint trait for the Miden Assembly AST:

* Unnecessary newlines in some situations
* Missing newlines in others
* Incorrect printing of the executable module's "begin" block as a procedure export
* Inconsistent handling of docstrings

This PR also adds a test that covers most interesting combinations of AST nodes - we can add to it over time to cover any other edge cases we may find.

### CODE REVIEW

**Review by bobbinth — APPROVED** [review]
Not a very deep review from me - but looks good! Thank you!

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,11 @@
 
 ## 0.14.0 TBD
 
+## 0.13.2 TBD
+
+#### Fixes
+
+- Fixed various issues with pretty printing of Miden Assembly
```
[the remark this answers is not in the document]
Comment (bitwalker) [reply CHANGELOG.md:9]: I'll reapply this to `main`, probably makes sense to ship it in 0.13.2 since it is fixing known bugs in the formatting, but I could go either way if you all have a preference.

File: assembly/src/ast/tests.rs [hunk 2]
```diff
@@ -1171,3 +1171,114 @@ fn assert_parsing_line_unexpected_token() {
...
+        drop
+    else
+        u32wrapping_add
+    end
+end
+
+begin
+    push.1
+    push.1
+    push.100
```
Comment (plafer) [inline assembly/src/ast/tests.rs:1275]: Constants get evaluated (i.e. we go from `DEFAULT_CONST` to `100`) - I don't think formatting should be doing that

File: assembly/src/ast/tests.rs [hunk 3]
```diff
@@ -1171,3 +1171,114 @@ fn assert_parsing_line_unexpected_token() {
...
+        push.1
+        while.true
+            dup.3
+            dup.3
+            u32wrapping_add3
+            swap.1
+            push.1
+            u32overflowing_sub
+            swap.1
+            movdn.3
```
Comment (plafer) [inline assembly/src/ast/tests.rs:1258]: Comments here got removed from the original source

File: assembly/src/ast/tests.rs [hunk 4]
```diff
@@ -1171,3 +1171,114 @@ fn assert_parsing_line_unexpected_token() {
...
+    let formatted = module.to_string();
+    let expected = "\
+#! module doc
+#!
+#! with spaces
+
+#! Perform `a + b`, `n` times
+#!
+#! with spaces
+proc.add_n_times
```
Comment (plafer) [inline assembly/src/ast/tests.rs:1243]: Similarly here, comments on the same line as code got removed

**Review by plafer — CHANGES_REQUESTED** [review]
From a preliminary review, I think there are behaviors to be fixed. Will do a more in-depth review once these are resolved

File: assembly/src/ast/tests.rs (hunk 2 above)
Comment (bitwalker) [reply assembly/src/ast/tests.rs:1275]: That isn't the formatter, the parser + semantic analysis is rewriting constants. The formatter is simply pretty printing the AST as-is.

[the remark this answers is not in the document]
Comment (bitwalker) [reply assembly/src/ast/tests.rs:1275]: Same as my other comment - the formatter is just printing the AST it is given.

We would need to rework the parser to better preserve the original syntax written by the user if we want to roundtrip perfectly. That's something we should probably do to aid in developing our `rustfmt`-like tooling, but that is a separate issue.

File: assembly/src/ast/tests.rs (hunk 3 above)
Comment (bitwalker) [reply assembly/src/ast/tests.rs:1258]: The parser does not preserve comments currently. That would be something we'd want to change if we rework the parser.

File: assembly/src/ast/tests.rs (hunk 4 above)
Comment (bitwalker) [reply assembly/src/ast/tests.rs:1243]: Same as previous comment.

**Review by plafer — APPROVED** [review]
LGTM!

I initially didn't properly understand what `test_roundtrip_formatting()` was testing, and why the discrepancies between the original source and the pretty printed string are expected. So I would add a comment there to explain that (similar to what you put in the comment responses).
